core/src/main/java/org/apache/iceberg/EnvironmentContext.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class EnvironmentContext {
+  public static final String ENGINE_NAME = "engine-name";
+  public static final String ENGINE_VERSION = "engine-version";
+
+  private EnvironmentContext() {}
+
+  private static final Map<String, String> PROPERTIES = Maps.newConcurrentMap();
+  private static final ThreadLocal<Map<String, Supplier<String>>> THREAD_LOCAL_PROPERTIES =
+      ThreadLocal.withInitial(Maps::newConcurrentMap);
```

Comment:
Why put these inside a `ThreadLocal`? Then these need to be defined for every thread. Instead, you could use the fact that the `Supplier` should return the correct value for any thread and just use a simple map.

Using a `Supplier` makes it so someone else has to handle the `ThreadLocal`.

Reply:
this will produce incorrect results without a ThreadLocal because different threads will just overwrite the value for a particular key. I've added a test that shows this.